### DISCUSSION TITLE
Taster / Schalter - Jarvis v3

### DISCUSSION THREAD

**Discussion by ma-gu-16** [Zefau/ioBroker.jarvis#1175] [Q&A] [answered]
Hi , 

ich habe nun auch mal v3 installiert. Mir ist das mit dem Button nicht ganz klar. 

Mein Verständnis ist: 
Taster = senden bei klick immer true oder immer false, je nach dem was man möchte. 
und ein Schalter = schaltet true und wenn bereits true dann false und wenn false dann wieder true. (wie du es bereits hast, entweder via abfrage des datenpunktes oder interner value).

Sehe ich das falsch? Evtl könnte man auch die Labels Taster/Schalter nutzen wie bei der Elektroinstallation. 

Wie konfiguriere ich den in der aktuellen Version einen Taster, der immer TRUE sendet wenn ich klicke, egal welchen State irgendetwas hat. 

Danke.

_Originally posted by @ma-gu-16 in https://github.com/Zefau/ioBroker.jarvis/issues/408#issuecomment-968117596_

**Comment by Zefau** [comment]
Grundsätzlich wird der aktuelle Wert des Datenpunkts invertiert und gesetzt. Also wenn der Datenpunkt auf `true` steht, wird `false` geschickt bzw. wenn der Datenpunkt auf `false` steht, dann wird `true` geschickt. Sowohl beim Schalter, als auch beim Taster. Der Unterschied zum Taster ist, dass nach dem Schaltvorgang erneut der Ausgangswert des Datenpunkts geschickt wird. Also wenn der Datenpunkt auf `true` steht, wird `false` geschickt und danach wieder auf `true` gestellt. Es wird beim Taster also 2x geschaltet.

Sofern immer ein bestimmter Wert geschickt werden soll, muss in den Datenpunkt Eigenschaften `setOn` oder `setOff` genutzt werden. Steht also der Datenpunkt auf `true` wird der Wert aus `setOff` geschickt. Ist der Datenpunkt-Wert auf `false` dann wird `setOn` geschickt. Damit kann das Schaltverhalten also individualisiert werden.

Wenn also immer `true` geschickt werden soll unabhängig im Wert des Datenpunkts, dann muss wohl `setOn` als auch `setOff` in den Datenpunkt-Eigenschaften mit `true` eingestellt werden.

**Comment by ma-gu-16** [comment]
Ok Danke dir. das klappt, habe es mit on / off versucht, das scheint nciht zu gehen mit setOff und setOn klappts. 

Denke die 2 sollte man auch noch bei dem kleinen info eintragen. 

Bei dem Taster wird da auch 2mal an iobroker gesendet oder nur in jarvis "zurückgestellt".

**Reply by Zefau** [reply]
Es wird 2x an ioBroker geschickt.

**Reply by ma-gu-16** [reply]
dann geht ja alles an und kurz darauf wieder aus ?  müsste man dann im iobroker ein onchange (only true) machen, damit das klappt. ist das so angedacht?

**Reply by Zefau** [reply]
> dann geht ja alles an und kurz darauf wieder aus ? 

Alles? Halt der eine Datenpunkt des Geräts. Ja, an und dann wieder aus - wie ein Taster halt.

> müsste man dann im iobroker ein onchange (only true) machen, damit das klappt. ist das so angedacht?

Wie das in ioBroker verarbeitet wird da bin ich raus. Ist ja abhängig vom Adapter. Wenn du das mit blockly verarbeitest, dann genügt ja ein Schalter oder nicht?

**Reply by ma-gu-16** [reply]
Hast du mir einen Anwendungsfall wo man das benötigt? Ich versuche gerade es zu verstehen ;-) 
Wieso soll man den etwas "anschalten" und kurzdarauf wieder "aus" , - selbst wenn man z.b. Treppenlichtfunktion macht nimmt man da einen Timeout, o.ä.  Soll keine Kritik sein, interessiert mich nur, evtl. kann ich die Funktion ja gebrauchen :-D

**Reply by Zefau** [reply]
Siehe https://github.com/Zefau/ioBroker.jarvis/issues/757#issuecomment-918961755 und https://github.com/Zefau/ioBroker.jarvis/issues/1088

**Reply by ma-gu-16** [reply]
Ok, got it. 2 verschiedenen atenpunte für Trigger und Anzeige, dann macht das ggfl. Sinn.  Wenn man ein und den selben Datenpunkt nutzt , was bei mir meistens der Fall ist, macht das komplett keinen Sinn mehr. Aber nun is klar. Danke dir ;-)

**Comment by ma-gu-16** [comment]
Und nochmal :-( Bin nicht sicher ob ich etwas falsch mache oder es ein Bug ist: 

Ich möchte 2 Buttons haben. der Eine ist Optisch ein OFF Button und bei jedem klickt Sender er false, egal welcher Zustand von wem auch immer. 
Der 2. Button soll einfach true senden, immer. 
```
  "07d412b0-56ab-461e-899b-969075ffdddf": {
    "name": "aktion_licht_an",
    "icon": "",
    "label": "",
    "function": "user",
    "states": {
      "state": {
        "showState": false,
        "action": "scene.0.Licht.Lichter",
        "actionElement": "IconButtonAction",
        "icon": {
          "default": "lightbulb-group"
        },
        "label": "Alle Lichter einschalten",
        "state": "scene.0.Licht.Lichter",
        "bodyElement": null,
        "stateKey": "state",
        "display": "",
        "IconButtonActionConfig": {
          "flat": false,
          "pushButton": false,
          "iconTurnOff": "mdi-power"
        },
        "properties": {
          "initial": true,
          "setOn": true,
          "setOff": true
        }
      }
    },
    "options": {
      "suppressPopup": true
    },
    "attributes": {
      "_created": 1607018175598,
      "_updated": 1607018838179
    },
    "id": "07d412b0-56ab-461e-899b-969075ffdddf",
    "revision": "e0af2"
  },
  "ffdc1a9b-7aed-4dd4-bdcb-10228c51b242": {
    "name": "aktion_licht_aus",
    "icon": "",
    "label": "",
    "function": "user",
    "states": {
      "state": {
        "showState": false,
        "action": "scene.0.Licht.Lichter",
        "actionElement": "IconButtonAction",
        "ButtonActionConfig": {
          "label": "ausschalten"
        },
        "icon": {
          "default": "lightbulb-group-off",
          "true": "lightbulb-group-off",
          "false": "lightbulb-group-off"
        },
        "label": "Alle Lichter ausschalten",
        "bodyElement": null,
        "properties": {
          "initial": false,
          "setOn": false,
          "setOff": false
        },
        "state": "scene.0.Licht.Lichter",
        "stateKey": "state",
        "IconButtonActionConfig": {
          "iconTurnOff": "mdi-power",
          "pushButton": true,
          "flat": false
        }
      }
    },
    "options": {
      "suppressPopup": true
    },
    "attributes": {
      "_created": 1607018381172,
      "_updated": 1607217426885
    },
    "id": "ffdc1a9b-7aed-4dd4-bdcb-10228c51b242",
    "revision": "55d14"
  },
```

Auch scheint mir irgendwo eine Art Cache zu haben, wie du siehst ist das Label des 2. Buttons falsch. Müsste wie in der Config oben ja ""Alle Lichter ausschalten" heissen.  Im Layout habe ich zwar die möglich das nochmal zu überschreiben, aber irgendwas stimmt hier nicht. 

![image](https://user-images.githubusercontent.com/22169074/142034311-4e49427b-af31-4fd6-9b24-8aad4657b7bf.png)

**Comment by ma-gu-16** [comment]
Hat jemand eine Lösung dafür ?  Bisher konnte ich das nicht ordentlich umsetzten :-(

**Comment by Bernd9000** [comment]
Ich mache es immer noch mit der alten Methode .
IconButtonAction und dann folgendes unter Datenpunkt Eigenschaften (JSON Format) eintragen:
Erster Schalter : {"on":"false","off":"false"}
Zweiter Schalter : {"on":"true","off":"true"}
Das sollte eigentlich funktionieren

**Reply by ma-gu-16** [reply]
Kannst du mal deine JSON posten von einem der buttons ? 

Danke dir.

**Comment by Bernd9000** [comment]
Ups, Du hast Recht.
Wenn für beide Schalter der selbe Datenpunkt benutzt wird funktioniert das nicht mehr.
Bei getrennten Datenpunkten funktioniert es.

**Reply by ma-gu-16** [reply]
genau das meinte ich :-(

**Comment by Bernd9000** [comment]
Mit "initial" "setOn" "setOff"  klappt es auch nicht (wenn beide Schalter den selben Datenpunkt schalten)
Da mußt Du wohl auf ein script zurückgreifen.

**Comment by Bernd9000** [comment]
Also, ich hab jetzt zwei Schalter.
Unter Datenpunkt Eigenschaften steht folgendes:
1.) Schalter = {"setOn":true,"setOff":true}
2.) Schalter = {"setOn":false,"setOff":false}

Wenn die beiden Schalter verschiedene Datenpunkte benutzen ist alles Okay.

Benutzen Sie den selben Datepunkt funktioniert der 1.Schalter richtig.
Der zweite Schalter sendet aber nur noch true.

Dann habe ich die Datenpunkteigenschaften mal getauscht.
1.) Schalter = {"setOn":false,"setOff":false}
2.) Schalter = {"setOn":true,"setOff":true}

Dann funktioniert wieder nur der erste Schalter.
Der zweite sendet nur noch false.

**Comment by ma-gu-16** [comment]
Denke der Workaround ist ich lege irgendein dummy object/datenpunkt mit false an und setzte den bei allen tastern als anzeigedatenpunkt (read-only).

und dann das setOn/setOff ... werde es heute abend mal versuchen.
Vorteil ist sicher das das Icon immer auf OFF bleibt.

@Zefau
Ein "einfacher" Button oder Taster (ohne Wechselwirkung) wäre sicher von Vorteil.
Bzw für mein Verständnis ist das der Button. One-Way wert senden und ggfl Icon nach dem Wert des Anzeigedatepunkts wechseln.
Diese zusätzliche interne Logik macht es irgendwie kompliziert.

Beim Taster ist das ja okay, das man dort ggfl zwischen interner logik und datenpunktvalue switchen kann um nach dem ON direkt OFF betätigen zu können.

Anwendungsfall:

Ein Button -> sendet true -> alle Lichter an.
Ein Button -> sendet false -> alle Lichter aus.

Dahinter sind jeweils diverse Scenen oder Scripte die auf Datenpunkte reagieren.

**Comment by ma-gu-16** [comment]
@Zefau Ist dies den möglich? Kannst du ggfl eine Beispiel Config posten ?

**Comment by ma-gu-16** [comment, marked as answer]
klappt, siehe: https://github.com/Zefau/ioBroker.jarvis/issues/1195

**Comment by ma-gu-16** [comment]
@Bernd9000 @Zefau 

Folgendes: 

Ich möchte eine Scene im iobroker schalten, true/false. 

finde keine passende Einstellung damit das sauber funktioniert. Weder Taster noch Button noch SwitchAction. 
über Objekte im iob steht der wert korrekt und ich kann ihn auch setzten true/false. 

Ziel: 
Button oder Switch schaltet false wenn iob state true ist und umgekehrt. Das sollte doch Default sein oder irre ich mich ? 


Habt ihr sowas umgesetzt? Ich weiss auch nicht ob es ein Bug ist oder Konfiguration. Ich denke eine ausführliche Anleitung wäre sicher hilfreich. 

Danke

**Comment by Bernd9000** [comment]
@ma-gu-16
Wenn der Schalter/Taster den aktuellen Datenpunkt auswerten soll, darfst Du nicht setOn oder setOff benutzen.
Hier ein ganz normaler Taster der den Datenpunkt vom Typ boolean auf true setzt wenn er false war und umgekehrt.

Datenpunkt Eigenschaften :   {"on":true,"off":false}

Wenn der Datenpunkt vom Datentyp String sein soll dann in den Datenpunkt Eigenschaften true und false in Anführungszeichen setzen.
Also so :   Datenpunkt Eigenschaften :   {"on":"true","off":"false"}

![Unbenannt](https://user-images.githubusercontent.com/77005685/145207687-2d143821-4b38-4b22-bf4e-51f1d5af3c66.jpg)

**Comment by ma-gu-16** [comment]
OMG, es klappt. 

Problem war das im Widget nochmal die Button Action augewählt war und somit meine Einstellungen von dem Gerät überschrieben hat. 
Wenn ich dort nichts auswähle müsste dann nicht die Default einstellung vom Gerät kommen ? 
Bei mir kommt da nichts. 

Das Trigger-Element sollte ja leer sein und dann sollte das vom Gerät genommen werden, oder nicht ? 

![image](https://user-images.githubusercontent.com/22169074/145257925-b83eb9aa-a41c-486b-9a08-561bc5742b5d.png)

**Reply by Bernd9000** [reply]
Ja, hatte ich auch so verstanden.
Wenn im Widget unter Trigger Element nichts angegeben wurde wird das Action Element vom Gerät als Rückfallebene benutzt.
Geht bei mir aber auch nicht mehr, wenn ich das Trigger Element im Widget lösche wird nichts mehr angezeigt.
